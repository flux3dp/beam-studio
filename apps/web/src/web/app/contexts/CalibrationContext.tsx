import React, { createContext, useState } from 'react';

import DeviceMaster from 'helpers/device-master';
import PreviewModeController from 'app/actions/beambox/preview-mode-controller';
import storage from 'implementations/storage';
import versionChecker from 'helpers/version-checker';
import { CameraConfig } from 'interfaces/Camera';
import {
  DEFAULT_CAMERA_OFFSET,
  STEP_ASK_READJUST,
  STEP_PUT_PAPER,
} from 'app/constants/camera-calibration-constants';
import { IDeviceInfo } from 'interfaces/IDevice';

interface CalibrationContextType {
  borderless: boolean;
  cameraPosition: { x: number; y: number };
  setCameraPosition: (val: { x: number; y: number }) => void;
  calibratedMachines: string[];
  setCalibratedMachines: (val: string[]) => void;
  currentOffset: CameraConfig;
  setCurrentOffset: (val: CameraConfig) => void;
  currentStep: symbol;
  setCurrentStep: (val: symbol) => void;
  lastConfig: CameraConfig;
  setLastConfig: (val: CameraConfig) => void;
  gotoNextStep: (step: symbol) => void;
  onClose: (completed: boolean) => void;
  originFanSpeed: number;
  setOriginFanSpeed: (val: number) => void;
  imgBlobUrl: string;
  setImgBlobUrl: (v: string) => void;
  device: IDeviceInfo;
  unit: string;
}

export const CalibrationContext = createContext<CalibrationContextType>({
  borderless: false,
  cameraPosition: { x: 0, y: 0 },
  setCameraPosition: () => {},
  calibratedMachines: [],
  setCalibratedMachines: () => {},
  currentOffset: null,
  setCurrentOffset: () => {},
  currentStep: STEP_PUT_PAPER,
  setCurrentStep: () => {},
  lastConfig: null,
  setLastConfig: () => {},
  gotoNextStep: () => {},
  onClose: () => {},
  imgBlobUrl: '',
  setImgBlobUrl: () => {},
  originFanSpeed: 100,
  setOriginFanSpeed: () => {},
  device: null,
  unit: 'mm',
});

interface CalibrationProviderProps {
  children: React.ReactNode;
  borderless: boolean;
  device: IDeviceInfo;
  onClose: (completed: boolean) => void;
}

let calibratedMachines = [];

export function CalibrationProvider({
  children,
  borderless,
  device,
  onClose,
}: CalibrationProviderProps): JSX.Element {
  const didCalibrate = calibratedMachines.includes(device.uuid);
  const initStep: symbol = didCalibrate ? STEP_ASK_READJUST : STEP_PUT_PAPER;
  const [currentStep, setCurrentStep] = useState(initStep);
  const [currentOffset, setCurrentOffset] = useState(DEFAULT_CAMERA_OFFSET);
  const [lastConfig, setLastConfig] = useState(DEFAULT_CAMERA_OFFSET);
  const setCalibratedMachines = (newList: string[]) => {
    calibratedMachines = newList;
  };
  const [imgBlobUrl, setImgBlobUrl] = useState('');
  const [cameraPosition, setCameraPosition] = useState({ x: 0, y: 0 });
  const [originFanSpeed, setOriginFanSpeed] = useState(1000);
  const unit = (storage.get('default-units') as string) || 'mm';

  const wrappedOnClose = async (completed: boolean) => {
    onClose(completed);
    await PreviewModeController.end({ shouldWaitForEnd: true });
    const tempCmdAvailable = versionChecker(device?.version).meetRequirement('TEMP_I2C_CMD');
    if (originFanSpeed && !tempCmdAvailable) {
      await DeviceMaster.setFan(originFanSpeed);
    }
  };

  return (
    <CalibrationContext.Provider
      value={{
        borderless,
        device,
        currentStep,
        setCurrentStep,
        cameraPosition,
        setCameraPosition,
        currentOffset,
        setCurrentOffset,
        lastConfig,
        setLastConfig,
        calibratedMachines,
        setCalibratedMachines,
        imgBlobUrl,
        setImgBlobUrl,
        originFanSpeed,
        setOriginFanSpeed,
        gotoNextStep: setCurrentStep,
        unit,
        onClose: wrappedOnClose,
      }}
    >
      {children}
    </CalibrationContext.Provider>
  );
}
