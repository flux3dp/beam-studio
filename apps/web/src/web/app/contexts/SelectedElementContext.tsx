import React, { createContext, memo, useCallback, useEffect, useState, useRef } from 'react';

import eventEmitterFactory from 'helpers/eventEmitterFactory';

const canvasEventEmitter = eventEmitterFactory.createEventEmitter('canvas');

export interface SelectedElementContextType {
  selectedElement: Element | null;
}

export const SelectedElementContext = createContext<SelectedElementContextType>({
  selectedElement: null,
});

interface Props {
  children: React.ReactNode;
}

export const SelectedElementContextProvider = memo(({ children }: Props): JSX.Element => {
  const [selectedElement, setSelectedElement] = useState<Element | null>(null);
  const selectedElementRef = useRef<Element | null>(null);

  const handleSetSelectedElem = useCallback(
    (elem: Element): void => {
      if (elem === selectedElementRef.current) return;
      selectedElementRef.current = elem;
      setSelectedElement((cur) => {
        if (cur === elem) return cur;
        (document.activeElement as HTMLInputElement).blur();
        return elem;
      });
    },
    []
  );

  useEffect(() => {
    canvasEventEmitter.on('SET_SELECTED_ELEMENT', handleSetSelectedElem);
    return () => {
      canvasEventEmitter.off('SET_SELECTED_ELEMENT', handleSetSelectedElem);
    };
  }, [handleSetSelectedElem]);

  return (
    <SelectedElementContext.Provider value={{ selectedElement }}>
      {children}
    </SelectedElementContext.Provider>
  );
});

export default {
  SelectedElementContext,
  SelectedElementContextProvider,
};
