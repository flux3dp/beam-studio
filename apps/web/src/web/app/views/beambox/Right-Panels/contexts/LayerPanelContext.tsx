import React, { createContext, useEffect, useState, useCallback, useMemo } from 'react';

import doLayersContainsVector from 'helpers/layer/check-vector';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import useForceUpdate from 'helpers/use-force-update';
import useWorkarea from 'helpers/hooks/useWorkarea';
import { getLayerElementByName } from 'helpers/layer/layer-helper';
import { promarkModels } from 'app/actions/beambox/constant';

interface ILayerPanelContext {
  selectedLayers: string[];
  setSelectedLayers: (selectedLayers: string[]) => void;
  forceUpdate: () => void;
  forceUpdateSelectedLayers: () => void;
  hasVector: boolean;
  hasGradient: boolean;
}

export const LayerPanelContext = createContext<ILayerPanelContext>({
  selectedLayers: [],
  setSelectedLayers: () => {},
  forceUpdate: () => {},
  forceUpdateSelectedLayers: () => {},
  hasVector: false,
  hasGradient: false,
});
const layerPanelEventEmitter = eventEmitterFactory.createEventEmitter('layer-panel');

interface Props {
  children?: React.ReactNode;
}

export const LayerPanelContextProvider = ({ children }: Props): JSX.Element => {
  const [hasVector, setHasVector] = useState<boolean>(false);
  const [hasGradient, setHasGradient] = useState<boolean>(false);
  const [selectedLayers, setSelectedLayers] = useState<string[]>([]);
  const forceUpdate = useForceUpdate();
  const forceUpdateSelectedLayers = useCallback(
    () => setSelectedLayers([...selectedLayers]),
    [selectedLayers]
  );
  const workarea = useWorkarea();
  const isPromark = useMemo(() => promarkModels.has(workarea), [workarea]);

  const lazySetSelectedLayers = useCallback(
    (newLayers: string[]) => {
      if (
        newLayers.length === selectedLayers.length &&
        newLayers.every((name, i) => name === selectedLayers[i])
      ) {
        return;
      }
      setSelectedLayers(newLayers);
    },
    [selectedLayers, setSelectedLayers]
  );

  useEffect(() => {
    layerPanelEventEmitter.on('UPDATE_LAYER_PANEL', forceUpdate);
    return () => {
      layerPanelEventEmitter.removeListener('UPDATE_LAYER_PANEL', forceUpdate);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    layerPanelEventEmitter.on('SET_SELECTED_LAYERS', lazySetSelectedLayers);
    return () => {
      layerPanelEventEmitter.removeListener('SET_SELECTED_LAYERS', lazySetSelectedLayers);
    };
  }, [lazySetSelectedLayers]);

  useEffect(() => {
    const getSelectedLayers = (response: { selectedLayers: string[] }) => {
      response.selectedLayers = selectedLayers;
    };
    layerPanelEventEmitter.on('GET_SELECTED_LAYERS', getSelectedLayers);
    const checkVector = () => {
      const newVal = doLayersContainsVector(selectedLayers);
      setHasVector(newVal);
    };
    layerPanelEventEmitter.on('CHECK_VECTOR', checkVector);
    return () => {
      layerPanelEventEmitter.removeListener('GET_SELECTED_LAYERS', getSelectedLayers);
      layerPanelEventEmitter.removeListener('CHECK_VECTOR', checkVector);
    };
  }, [selectedLayers]);

  useEffect(() => {
    const checkGradient = () => {
      if (isPromark) {
        const newVal = selectedLayers.some((layerName: string) => {
          const layer = getLayerElementByName(layerName);
          return !!layer?.querySelector('image[data-shading="true"]');
        });
        setHasGradient(newVal);
      }
    };
    checkGradient();
    layerPanelEventEmitter.on('CHECK_GRADIENT', checkGradient);
    return () => {
      layerPanelEventEmitter.removeListener('CHECK_GRADIENT', checkGradient);
    };
  }, [selectedLayers, isPromark]);

  // move doLayersContainsVector and setVector in effect to avoid heavy render process
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    const newVal = doLayersContainsVector(selectedLayers);
    if (hasVector !== newVal) {
      setHasVector(newVal);
    }
    if (isPromark) {
      const newGradientVal = selectedLayers.some((layerName: string) => {
        const layer = getLayerElementByName(layerName);
        return !!layer?.querySelector('image[data-shading="true"]');
      });
      if (newGradientVal !== hasGradient) setHasGradient(newGradientVal);
    }
  });

  return (
    <LayerPanelContext.Provider
      value={{
        selectedLayers,
        setSelectedLayers: lazySetSelectedLayers,
        forceUpdate,
        forceUpdateSelectedLayers,
        hasVector,
        hasGradient,
      }}
    >
      {children}
    </LayerPanelContext.Provider>
  );
};
