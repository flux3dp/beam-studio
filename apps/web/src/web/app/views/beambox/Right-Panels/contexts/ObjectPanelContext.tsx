import React, { useCallback, useEffect, useRef, useState } from 'react';

import eventEmitterFactory from 'helpers/eventEmitterFactory';
import useForceUpdate from 'helpers/use-force-update';

interface IObjectPanelContext {
  polygonSides: number;
  dimensionValues: any;
  activeKey: string | null;
  updateDimensionValues: (newValues: any) => void;
  getDimensionValues: (
    response: {
      dimensionValues: any;
    },
    key?: string
  ) => any;
  updateActiveKey: (activeKey: string | null) => void;
  updateObjectPanel: () => void;
}

export const ObjectPanelContext = React.createContext<IObjectPanelContext>({
  polygonSides: 5,
  dimensionValues: {},
  activeKey: null,
  updateDimensionValues: () => {},
  getDimensionValues: () => {},
  updateActiveKey: () => {},
  updateObjectPanel: () => {},
});
const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');
const minRenderInterval = 200;

interface Props {
  children?: React.ReactNode;
}

export const ObjectPanelContextProvider = ({ children }: Props): JSX.Element => {
  const forceUpdate = useForceUpdate();
  const [polygonSides, setPolygonSides] = useState(5);
  const [activeKey, setActiveKey] = useState<string | null>(null);
  const dimensionValues = useRef<{ [key: string]: number | string }>({});
  const lastUpdateTime = useRef(Date.now());
  const updateTimeout = useRef<NodeJS.Timeout>(null);
  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_POLYGON_SIDES', setPolygonSides);
    objectPanelEventEmitter.on('UPDATE_ACTIVE_KEY', setActiveKey);
    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_POLYGON_SIDES', setPolygonSides);
      objectPanelEventEmitter.removeListener('UPDATE_ACTIVE_KEY', setActiveKey);
    }
  }, []);

  const getDimensionValues = useCallback((response: { dimensionValues: any }, key?: string) => {
    response.dimensionValues = key ? dimensionValues.current[key] : dimensionValues.current;
  }, []);

  useEffect(() => {
    objectPanelEventEmitter.on('GET_DIMENSION_VALUES', getDimensionValues);
    return () => {
      objectPanelEventEmitter.removeListener('GET_DIMENSION_VALUES', getDimensionValues);
    }
  }, [getDimensionValues]);
  const updateDimensionValues = useCallback((newValues: { [key: string]: number | string }) => {
    dimensionValues.current = {
      ...dimensionValues.current,
      ...newValues,
    };
  }, []);

  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_DIMENSION_VALUES', updateDimensionValues);
    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_DIMENSION_VALUES', updateDimensionValues);
    }
  }, [updateDimensionValues]);
  const getActiveKey = useCallback((response: { activeKey: string | null }) => {
    response.activeKey = activeKey;
  }, [activeKey]);

  useEffect(() => {
    objectPanelEventEmitter.on('GET_ACTIVE_KEY', getActiveKey);
    return () => {
      objectPanelEventEmitter.removeListener('GET_ACTIVE_KEY', getActiveKey);
    }
  }, [getActiveKey]);

  const updateObjectPanel = useCallback(() => {
    clearTimeout(updateTimeout.current);
    const time = Date.now();
    if (time - lastUpdateTime.current >= minRenderInterval) {
      lastUpdateTime.current = time;
      forceUpdate();
    } else {
      updateTimeout.current = setTimeout(() => {
        lastUpdateTime.current += minRenderInterval;
        forceUpdate();
      }, lastUpdateTime.current + minRenderInterval - time);
    }
  }, [forceUpdate]);
  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_OBJECT_PANEL', updateObjectPanel);
    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_OBJECT_PANEL', updateObjectPanel);
    }
  }, [updateObjectPanel]);

  return (
    <ObjectPanelContext.Provider
      value={{
        polygonSides,
        dimensionValues: dimensionValues.current,
        activeKey,
        updateDimensionValues,
        getDimensionValues,
        updateActiveKey: setActiveKey,
        updateObjectPanel,
      }}
    >
      {children}
    </ObjectPanelContext.Provider>
  );
};
