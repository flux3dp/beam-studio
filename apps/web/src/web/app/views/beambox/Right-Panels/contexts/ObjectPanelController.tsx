import eventEmitterFactory from 'helpers/eventEmitterFactory';

const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');

const updateDimensionValues = (newValue: any): void => {
  objectPanelEventEmitter.emit('UPDATE_DIMENSION_VALUES', newValue);
};

const getDimensionValues = (key?: string): any => {
  const response = {
    dimensionValues: {},
  };
  objectPanelEventEmitter.emit('GET_DIMENSION_VALUES', response, key);
  return response.dimensionValues;
};

const minEventInterval = 50;
let updateObjectPanelTimeout = null;
const updateObjectPanel = (): void => {
  clearTimeout(updateObjectPanelTimeout);
  updateObjectPanelTimeout = setTimeout(() => {
    objectPanelEventEmitter.emit('UPDATE_OBJECT_PANEL');
  }, minEventInterval);
};

const updatePolygonSides = (polygonSides: number): void => {
  objectPanelEventEmitter.emit('UPDATE_POLYGON_SIDES', polygonSides);
};

const updateActiveKey = (activeKey: string | null): void => {
  objectPanelEventEmitter.emit('UPDATE_ACTIVE_KEY', activeKey);
};

const getActiveKey = (): string | null => {
  const response = {
    activeKey: undefined,
  };
  objectPanelEventEmitter.emit('GET_ACTIVE_KEY', response);
  return response.activeKey;
};

export default {
  updateObjectPanel,
  updateDimensionValues,
  getDimensionValues,
  updatePolygonSides,
  updateActiveKey,
  getActiveKey,
  events: objectPanelEventEmitter,
};
