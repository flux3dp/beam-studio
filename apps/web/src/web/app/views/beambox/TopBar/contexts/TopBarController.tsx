import BeamboxGlobalInteraction from 'app/actions/beambox/beambox-global-interaction';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import { IDeviceInfo } from 'interfaces/IDevice';

const topBarEventEmitter = eventEmitterFactory.createEventEmitter('top-bar');

const setElement = (elem: Element | null): void => {
  if (!elem) {
    BeamboxGlobalInteraction.onObjectBlur();
  } else {
    BeamboxGlobalInteraction.onObjectBlur();
    BeamboxGlobalInteraction.onObjectFocus([elem]);
  }

  topBarEventEmitter.emit('SET_ELEMENT', elem);
};

const updateTitle = (title: string, isCloudFile: boolean): void => {
  topBarEventEmitter.emit('UPDATE_TITLE', title, isCloudFile);
};

const onTitleChange = (handler: (title: string, isCloudFile: boolean) => void): void => {
  topBarEventEmitter.on('UPDATE_TITLE', handler);
};

const offTitleChange = (handler: (title: string, isCloudFile: boolean) => void): void => {
  topBarEventEmitter.removeListener('UPDATE_TITLE', handler);
};

const setHasUnsavedChange = (hasUnsavedChange: boolean): void => {
  topBarEventEmitter.emit('SET_HAS_UNSAVED_CHANGE', hasUnsavedChange);
};

const getTopBarPreviewMode = (): boolean => {
  const response = {
    isPreviewMode: false,
  };
  topBarEventEmitter.emit('GET_TOP_BAR_PREVIEW_MODE', response);
  return response.isPreviewMode;
};

const getSelectedDevice = (): IDeviceInfo | null => {
  const response = {
    selectedDevice: null,
  };
  topBarEventEmitter.emit('GET_SELECTED_DEVICE', response);
  return response.selectedDevice;
};

const setSelectedDevice = (device: IDeviceInfo): void => {
  topBarEventEmitter.emit('SET_SELECTED_DEVICE', device);
};

export default {
  setElement,
  updateTitle,
  setHasUnsavedChange,
  getTopBarPreviewMode,
  getSelectedDevice,
  setSelectedDevice,
  onTitleChange,
  offTitleChange,
};
