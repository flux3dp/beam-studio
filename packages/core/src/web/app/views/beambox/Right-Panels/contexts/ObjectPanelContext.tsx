/* eslint-disable reactRefresh/only-export-components */
import React, { useCallback, useEffect, useRef, useState } from 'react';

import eventEmitterFactory from '@core/helpers/eventEmitterFactory';
import useForceUpdate from '@core/helpers/use-force-update';
import type { DimensionKey, DimensionValues } from '@core/interfaces/ObjectPanel';

interface IObjectPanelContext {
  activeKey: null | string;
  dimensionValues: any;
  getDimensionValues: (
    response: {
      dimensionValues: any;
    },
    key?: DimensionKey,
  ) => void;
  polygonSides: number;
  updateActiveKey: (activeKey: null | string) => void;
  updateDimensionValues: (newValues: DimensionValues) => void;
  updateObjectPanel: () => void;
}

export const ObjectPanelContext = React.createContext<IObjectPanelContext>({
  activeKey: null,
  dimensionValues: {},
  getDimensionValues: () => {},
  polygonSides: 5,
  updateActiveKey: () => {},
  updateDimensionValues: () => {},
  updateObjectPanel: () => {},
});

const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');
const minRenderInterval = 200;

interface Props {
  children?: React.ReactNode;
}

export const ObjectPanelContextProvider = ({ children }: Props): React.JSX.Element => {
  const forceUpdate = useForceUpdate();
  const [polygonSides, setPolygonSides] = useState(5);
  const [activeKey, setActiveKey] = useState<null | string>(null);
  const dimensionValues = useRef<DimensionValues>({});
  const lastUpdateTime = useRef(Date.now());
  const updateTimeout = useRef<NodeJS.Timeout | null>(null);

  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_POLYGON_SIDES', setPolygonSides);
    objectPanelEventEmitter.on('UPDATE_ACTIVE_KEY', setActiveKey);

    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_POLYGON_SIDES', setPolygonSides);
      objectPanelEventEmitter.removeListener('UPDATE_ACTIVE_KEY', setActiveKey);
    };
  }, []);

  const getDimensionValues = useCallback((response: { dimensionValues: any }, key?: DimensionKey) => {
    response.dimensionValues = key ? dimensionValues.current[key] : dimensionValues.current;
  }, []);

  useEffect(() => {
    objectPanelEventEmitter.on('GET_DIMENSION_VALUES', getDimensionValues);

    return () => {
      objectPanelEventEmitter.removeListener('GET_DIMENSION_VALUES', getDimensionValues);
    };
  }, [getDimensionValues]);

  const updateDimensionValues = useCallback((newValues: DimensionValues) => {
    dimensionValues.current = {
      ...dimensionValues.current,
      ...newValues,
    };
  }, []);

  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_DIMENSION_VALUES', updateDimensionValues);

    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_DIMENSION_VALUES', updateDimensionValues);
    };
  }, [updateDimensionValues]);

  const getActiveKey = useCallback(
    (response: { activeKey: null | string }) => {
      response.activeKey = activeKey;
    },
    [activeKey],
  );

  useEffect(() => {
    objectPanelEventEmitter.on('GET_ACTIVE_KEY', getActiveKey);

    return () => {
      objectPanelEventEmitter.removeListener('GET_ACTIVE_KEY', getActiveKey);
    };
  }, [getActiveKey]);

  const updateObjectPanel = useCallback(() => {
    if (updateTimeout.current) clearTimeout(updateTimeout.current);

    const time = Date.now();

    if (time - lastUpdateTime.current >= minRenderInterval) {
      lastUpdateTime.current = time;
      forceUpdate();
    } else {
      updateTimeout.current = setTimeout(
        () => {
          lastUpdateTime.current += minRenderInterval;
          forceUpdate();
        },
        lastUpdateTime.current + minRenderInterval - time,
      );
    }
  }, [forceUpdate]);

  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_OBJECT_PANEL', updateObjectPanel);

    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_OBJECT_PANEL', updateObjectPanel);
    };
  }, [updateObjectPanel]);

  return (
    <ObjectPanelContext.Provider
      value={{
        activeKey,
        dimensionValues: dimensionValues.current,
        getDimensionValues,
        polygonSides,
        updateActiveKey: setActiveKey,
        updateDimensionValues,
        updateObjectPanel,
      }}
    >
      {children}
    </ObjectPanelContext.Provider>
  );
};
