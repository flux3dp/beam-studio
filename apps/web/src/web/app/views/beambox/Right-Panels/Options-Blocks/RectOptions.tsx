import React from 'react';

import Constant from 'app/actions/beambox/constant';
import i18n from 'helpers/i18n';
import ObjectPanelController from 'app/views/beambox/Right-Panels/contexts/ObjectPanelController';
import ObjectPanelItem from 'app/views/beambox/Right-Panels/ObjectPanelItem';
import OptionPanelIcons from 'app/icons/option-panel/OptionPanelIcons';
import storage from 'implementations/storage';
import UnitInput from 'app/widgets/Unit-Input-v2';
import { getSVGAsync } from 'helpers/svg-editor-helper';
import { useIsMobile } from 'helpers/system-helper';

import styles from './RectOptions.module.scss';

let svgCanvas;
getSVGAsync((globalSVG) => {
  svgCanvas = globalSVG.Canvas;
});

const LANG = i18n.lang.beambox.right_panel.object_panel.option_panel;

interface Props {
  elem: Element;
  rx: number;
  updateDimensionValues: (val: { [key: string]: any }) => void;
}

function RectOptions({ elem, rx, updateDimensionValues }: Props): JSX.Element {
  const isMobile = useIsMobile();
  const handleRoundedCornerChange = (val) => {
    // eslint-disable-next-line no-param-reassign
    val *= Constant.dpmm;
    svgCanvas.changeSelectedAttribute('rx', val, [elem]);
    updateDimensionValues({ rx: val });
  };

  const renderRoundCornerBlock = () => {
    const unit = storage.get('default-units') || 'mm';
    const isInch = unit === 'inches';
    return isMobile ? (
      <ObjectPanelItem.Number
        id="rounded-corner"
        value={rx / Constant.dpmm || 0}
        min={0}
        updateValue={(val) => {
          handleRoundedCornerChange(val);
          ObjectPanelController.updateObjectPanel();
        }}
        label={LANG.rounded_corner}
      />
    ) : (
      <div className={styles['rounded-corner']} key="rounded-corner">
        <div className={styles.label} title={LANG.rounded_corner}>
          <OptionPanelIcons.RoundedCorner />
        </div>
        <UnitInput
          min={0}
          unit={isInch ? 'in' : 'mm'}
          className={{ 'option-input': true }}
          defaultValue={rx / Constant.dpmm || 0}
          getValue={(val) => handleRoundedCornerChange(val)}
        />
      </div>
    );
  };

  return isMobile ? renderRoundCornerBlock() : <div>{renderRoundCornerBlock()}</div>;
}

export default RectOptions;
