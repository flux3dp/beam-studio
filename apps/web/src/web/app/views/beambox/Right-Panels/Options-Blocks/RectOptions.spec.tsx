import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import { ObjectPanelContextProvider } from 'app/views/beambox/Right-Panels/contexts/ObjectPanelContext';

import RectOptions from './RectOptions';

jest.mock('app/views/beambox/Right-Panels/Options-Blocks/InFillBlock', () => () => <div>DummyInFillBlock</div>);

jest.mock('helpers/i18n', () => ({
  lang: {
    beambox: {
      right_panel: {
        object_panel: {
          option_panel: {
            rounded_corner: 'Rounded corner',
          },
        },
      },
    },
  },
}));

const get = jest.fn();
jest.mock('implementations/storage', () => ({
  get: (...args) => get(...args),
}));

const useIsMobile = jest.fn();
jest.mock('helpers/system-helper', () => ({
  useIsMobile: () => useIsMobile(),
}));

const changeSelectedAttribute = jest.fn();
jest.mock('helpers/svg-editor-helper', () => ({
  getSVGAsync: (callback) => {
    callback({
      Canvas: {
        changeSelectedAttribute: (...args) => changeSelectedAttribute(...args),
      },
    });
  },
}));

describe('should render correctly', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('unit is inches', () => {
    get.mockReturnValue('inches');
    const updateDimensionValues = jest.fn();
    const { container } = render(
      <RectOptions
        elem={document.getElementById('flux')}
        rx={0}
        updateDimensionValues={updateDimensionValues}
      />
    );

    expect(container).toMatchSnapshot();
    fireEvent.change(container.querySelector('input'), { target: { value: 1 } });
    fireEvent.blur(container.querySelector('input'));

    expect(changeSelectedAttribute).toHaveBeenCalledTimes(1);
    expect(changeSelectedAttribute).toHaveBeenNthCalledWith(1, 'rx', 254, [document.getElementById('flux')]);
    expect(updateDimensionValues).toHaveBeenCalledTimes(1);
    expect(updateDimensionValues).toHaveBeenNthCalledWith(1, { rx: 254 });
  });

  test('unit is not inches', () => {
    get.mockReturnValue(null);

    const { container } = render(
      <RectOptions
        elem={document.getElementById('flux')}
        rx={10}
        updateDimensionValues={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
});

describe('should render correctly in mobile', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('unit is inches', () => {
    useIsMobile.mockReturnValue(true);
    get.mockReturnValue('inches');
    const updateDimensionValues = jest.fn();
    const { baseElement, container } = render(
      <ObjectPanelContextProvider>
        <RectOptions
          elem={document.getElementById('flux')}
          rx={0}
          updateDimensionValues={updateDimensionValues}
        />
      </ObjectPanelContextProvider>
    );
    expect(container).toMatchSnapshot();

    const objectPanelItem = baseElement.querySelector('div.object-panel-item');
    fireEvent.click(objectPanelItem);
    fireEvent.click(baseElement.querySelectorAll('.step-buttons button')[1]);
    expect(changeSelectedAttribute).toHaveBeenCalledTimes(1);
    expect(changeSelectedAttribute).toHaveBeenNthCalledWith(1, 'rx', 254, [document.getElementById('flux')]);
    expect(updateDimensionValues).toHaveBeenCalledTimes(1);
    expect(updateDimensionValues).toHaveBeenNthCalledWith(1, { rx: 254 });
  });

  test('unit is not inches', () => {
    useIsMobile.mockReturnValue(true);
    get.mockReturnValue(null);
    const { container } = render(
      <RectOptions
        elem={document.getElementById('flux')}
        rx={10}
        updateDimensionValues={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
