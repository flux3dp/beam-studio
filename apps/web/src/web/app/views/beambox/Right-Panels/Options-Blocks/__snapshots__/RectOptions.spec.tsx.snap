// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly in mobile unit is inches 1`] = `
<div>
  <div
    class="object-panel-item"
    id="rounded-corner"
  >
    <div
      class="main"
    >
      <button
        class="adm-button adm-button-default adm-button-fill-outline adm-button-mini adm-button-shape-rounded number-item"
        type="button"
      >
        <span>
          0
        </span>
      </button>
    </div>
    <div
      class="label"
    >
      Rounded corner
    </div>
  </div>
</div>
`;

exports[`should render correctly in mobile unit is not inches 1`] = `
<div>
  <div
    class="object-panel-item"
    id="rounded-corner"
  >
    <div
      class="main"
    >
      <button
        class="adm-button adm-button-default adm-button-fill-outline adm-button-mini adm-button-shape-rounded number-item"
        type="button"
      >
        <span>
          1
        </span>
      </button>
    </div>
    <div
      class="label"
    >
      Rounded corner
    </div>
  </div>
</div>
`;

exports[`should render correctly unit is inches 1`] = `
<div>
  <div>
    <div
      class="rounded-corner"
    >
      <div
        class="label"
        title="Rounded corner"
      >
        <svgr-url />
      </div>
      <div
        class="option-input ui ui-control-unit-input-v2"
      >
        <input
          step="1"
          type="text"
          value="0.0000"
        />
        <span
          class="unit"
        >
          in
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly unit is not inches 1`] = `
<div>
  <div>
    <div
      class="rounded-corner"
    >
      <div
        class="label"
        title="Rounded corner"
      >
        <svgr-url />
      </div>
      <div
        class="option-input ui ui-control-unit-input-v2"
      >
        <input
          step="1"
          type="text"
          value="1.00"
        />
        <span
          class="unit"
        >
          mm
        </span>
      </div>
    </div>
  </div>
</div>
`;
